File: eval-reviews/2026-09-12-chained/query-null/typescript/sol/r1/starter/engine.ts
/** Stable relational operators and real constant-fold / scan-filter rewrites. */
import { aggregates } from "./model.ts";
import type { Expr, Plan, Value } from "./model.ts";
import { walk } from "./binder.ts";
export function evaluate(e: Expr, row: Value[], group: Value[][] = []): Value {
  if (e.op === "lit") return e.value as Value;
  if (e.op === "col") return row[e.index!];
  if (aggregates.has(e.op)) {
    if (e.op === "COUNT")
      return e.args.length
        ? group.reduce((n, r) => n + (evaluate(e.args[0], r) === null ? 0 : 1), 0)
        : group.length;
    const vs = group
      .map((r) => evaluate(e.args[0], r))
      .filter((v) => v !== null);
    if (!vs.length) return null;
    if (e.op === "SUM")
      return vs.reduce<number>((n, v) => n + (v as number), 0);
    return vs.reduce((a, b) =>
      e.op === "MIN" ? (a < b ? a : b) : a > b ? a : b,
    );
  }
  if (e.op === "COALESCE") {
    for (const arg of e.args) {
      const value = evaluate(arg, row, group);
      if (value !== null) return value;
    }
    return null;
  }
  const a = evaluate(e.args[0], row, group);
  if (e.op === "IS NULL") return a === null;
  if (e.op === "IS NOT NULL") return a !== null;
  if (e.op === "NOT") return a === null ? null : !a;
  const b = evaluate(e.args[1], row, group);
  if (e.op === "AND") {
    if (a === false || b === false) return false;
    return a === null || b === null ? null : true;
  }
  if (e.op === "OR") {
    if (a === true || b === true) return true;
    return a === null || b === null ? null : false;
  }
  if (a === null || b === null) return null;
  switch (e.op) {
    case "+":
      return (a as number) + (b as number);
    case "-":
      return (a as number) - (b as number);
    case "*":
      return (a as number) * (b as number);
    case "=":
      return a === b;
    case "<>":
      return a !== b;
    case "<":
      return a < b;
    case "<=":
      return a <= b;
    case ">":
      return a > b;
    case ">=":
      return a >= b;
    default:
      throw new Error(`Unknown bound operator ${e.op}`);
  }
}
function fold(e: Expr): Expr {
  e.args = e.args.map(fold);
  return !aggregates.has(e.op) &&
    !["lit", "col"].includes(e.op) &&
    e.args.every((a) => a.op === "lit")
    ? { op: "lit", value: evaluate(e, []), args: [], type: e.type }
    : e;
}
function conjuncts(e: Expr): Expr[] {
  return e.op === "AND" ? e.args.flatMap(conjuncts) : [e];
}
export function optimize(plan: Plan): Plan {
  const q = plan.query;
  q.select = q.select.map(([e, a]) => [fold(e), a]);
  q.joins = q.joins.map((j) => ({ ...j, on: fold(j.on) }));
  if (q.having) q.having = fold(q.having);
  if (q.where) {
    const remaining: Expr[] = [];
    for (const e of conjuncts(fold(q.where))) {
      const sources = new Set(
        [...walk(e)].filter((n) => n.op === "col").map((n) => n.source!),
      );
      if (
        plan.tables.length > 1 &&
        q.joins.every((j) => j.kind === "inner") &&
        sources.size === 1
      )
        plan.filters[[...sources][0]].push(e);
      else remaining.push(e);
    }
    q.where = remaining.reduce<Expr | undefined>(
      (a, e) => (a ? { op: "AND", args: [a, e], type: "bool" } : e),
      undefined,
    );
  }
  return plan;
}
export function execute(plan: Plan): { columns: string[]; rows: Value[][] } {
  const q = plan.query;
  let offset = 0;
  const scans = plan.tables.map((t, i) => {
    const rows = t.rows.filter((r) =>
      plan.filters[i].every((p) =>
        evaluate(p, [...Array<Value>(offset).fill(null), ...r]) === true,
      ),
    );
    offset += t.columns.length;
    return rows;
  });
  let rows = scans[0];
  q.joins.forEach((join, i) => {
    rows = rows.flatMap((left) => {
      const matches = scans[i + 1]
        .map((right) => [...left, ...right])
        .filter((row) => evaluate(join.on, row) === true);
      return matches.length || join.kind === "inner"
        ? matches
        : [[...left, ...Array<Value>(plan.tables[i + 1].columns.length).fill(null)]];
    });
  });
  if (q.where) rows = rows.filter((r) => evaluate(q.where!, r) === true);
  let units: [Value[], Value[][]][];
  if (plan.aggregate) {
    const groups = new Map<string, Value[][]>();
    for (const row of rows) {
      const key = JSON.stringify(q.groups.map((e) => evaluate(e, row)));
      const g = groups.get(key);
      if (g) g.push(row);
      else groups.set(key, [row]);
    }
    if (!q.groups.length && !rows.length) groups.set("[]", []);
    units = [...groups.values()].map((g) => [g[0] ?? [], g]);
  } else units = rows.map((r) => [r, []]);
  let projected = units
    .filter(([r, g]) => !q.having || evaluate(q.having, r, g) === true)
    .map(([r, g]) => q.select.map(([e]) => evaluate(e, r, g)));
  if (q.distinct) {
    const seen = new Set<string>();
    projected = projected.filter((r) => {
      const key = JSON.stringify(r);
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  }
  projected.sort((a, b) => {
    for (const [index, desc, explicitNullsFirst] of q.order) {
      const i = index as number;
      if (a[i] === null || b[i] === null) {
        if (a[i] === b[i]) continue;
        const nullsFirst = explicitNullsFirst ?? desc;
        return a[i] === null ? (nullsFirst ? -1 : 1) : nullsFirst ? 1 : -1;
      }
      const c = a[i] < b[i] ? -1 : a[i] > b[i] ? 1 : 0;
      if (c) return desc ? -c : c;
    }
    return 0;
  });
  return {
    columns: q.select.map(([, a]) => a),
    rows: projected.slice(
      q.offset,
      q.limit === undefined ? undefined : q.offset + q.limit,
    ),
  };
}

File: eval-reviews/2026-09-12-chained/query-null/python/sol/r1/starter/engine.py
"""Stable relational execution and semantics-preserving baseline optimization."""
from model import Expr
from parser import AGGREGATES
from binder import walk


def evaluate(e, row, group=None):
    if e.op == 'lit': return e.value
    if e.op == 'col': return row[e.index]
    if e.op in AGGREGATES:
        if e.op == 'COUNT':
            return len(group) if not e.args else sum(evaluate(e.args[0], r) is not None for r in group)
        vs = [evaluate(e.args[0], r) for r in group]
        vs = [v for v in vs if v is not None]
        if not vs: return None
        return {'SUM': sum, 'MIN': min, 'MAX': max}[e.op](vs)
    if e.op == 'COALESCE':
        for arg in e.args:
            value = evaluate(arg, row, group)
            if value is not None: return value
        return None
    a = evaluate(e.args[0], row, group)
    if e.op == 'IS_NULL': return a is None
    if e.op == 'IS_NOT_NULL': return a is not None
    if e.op == 'NOT': return None if a is None else not a
    b = evaluate(e.args[1], row, group)
    if e.op == 'AND':
        if a is False or b is False: return False
        if a is None or b is None: return None
        return True
    if e.op == 'OR':
        if a is True or b is True: return True
        if a is None or b is None: return None
        return False
    if a is None or b is None: return None
    if e.op == '+': return a + b
    if e.op == '-': return a - b
    if e.op == '*': return a * b
    if e.op == '=': return a == b
    if e.op == '<>': return a != b
    if e.op == '<': return a < b
    if e.op == '<=': return a <= b
    if e.op == '>': return a > b
    if e.op == '>=': return a >= b
    raise AssertionError(e.op)


def fold(e):
    e.args = [fold(a) for a in e.args]
    if e.op not in AGGREGATES | {'lit', 'col'} and all(a.op == 'lit' for a in e.args):
        return Expr('lit', evaluate(e, []), type=e.type)
    return e


def conjuncts(e):
    if e.op == 'AND': return conjuncts(e.args[0]) + conjuncts(e.args[1])
    return [e]


def optimize(plan):
    q = plan.query
    q.select = [(fold(e), a) for e, a in q.select]
    q.joins = [(kind, fold(e)) for kind, e in q.joins]
    if q.having: q.having = fold(q.having)
    if q.where:
        remaining = []
        for e in conjuncts(fold(q.where)):
            sources = {n.source for n in walk(e) if n.op == 'col'}
            if len(plan.tables) > 1 and all(kind == 'inner' for kind, _ in q.joins) and len(sources) == 1:
                plan.filters[next(iter(sources))].append(e)
            else: remaining.append(e)
        q.where = None
        for e in remaining:
            q.where = e if q.where is None else Expr('AND', args=[q.where, e], type='bool')
    return plan


def execute(plan):
    q = plan.query
    scans, offset = [], 0
    for t, predicates in zip(plan.tables, plan.filters):
        scans.append([r for r in t['rows'] if all(evaluate(p, [None] * offset + r) is True for p in predicates)])
        offset += len(t['columns'])
    rows = scans[0]
    for table, right, (kind, on) in zip(plan.tables[1:], scans[1:], q.joins):
        joined = []
        padding = [None] * len(table['columns'])
        for left in rows:
            matches = [left + r for r in right if evaluate(on, left + r) is True]
            joined.extend(matches if matches else ([left + padding] if kind == 'left' else []))
        rows = joined
    if q.where: rows = [r for r in rows if evaluate(q.where, r) is True]
    if plan.aggregate:
        groups = {}
        for row in rows:
            key = tuple(evaluate(e, row) for e in q.groups)
            groups.setdefault(key, []).append(row)
        if not q.groups and not rows: groups[()] = []
        units = [(g[0] if g else [], g) for g in groups.values()]
    else: units = [(r, None) for r in rows]
    projected = [[evaluate(e, r, g) for e, _ in q.select] for r, g in units if q.having is None or evaluate(q.having, r, g) is True]
    return format_projected(plan, projected)


def format_projected(plan, projected):
    """Apply the post-projection operators to cached view contributions."""
    q = plan.query
    projected = [list(r) for r in projected]
    if q.distinct: projected = [list(r) for r in dict.fromkeys(tuple(r) for r in projected)]
    for i, descending, explicit in reversed(q.order):
        null_first = explicit == 'first' if explicit else descending
        null_marker = 1 if (null_first == descending) else 0
        projected.sort(key=lambda r: (null_marker if r[i] is None else 1 - null_marker,
                                      0 if r[i] is None else r[i]), reverse=descending)
    projected = projected[q.offset: None if q.limit is None else q.offset + q.limit]
    return {'columns': [a for _, a in q.select], 'rows': projected}
